### PR TITLE
planner/core: support union all for mpp. (#24287)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #24287 to release-5.0
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/25051
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/25051:release-5.0-52e89cb8bfba
```

---

### What problem does this PR solve?
Problem Summary:
Support union all operator push down

### What is changed and how it works?

Proposal: [proposal of union all](https://github.com/pingcap/tics/issues/1829) 

What's Changed:
multiplying tasks during buidling mpp tasks.

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

### Release note 

- planner/core: support union all for mpp.

### CODE REVIEW

**Review by lzmhhh123 — APPROVED** [review]

File: expression/column.go [hunk 1]
```diff
@@ -38,10 +38,9 @@ type CorrelatedColumn struct {
 
 // Clone implements Expression interface.
 func (col *CorrelatedColumn) Clone() Expression {
-	var d types.Datum
 	return &CorrelatedColumn{
 		Column: col.Column,
-		Data:   &d,
+		Data:   col.Data,
```
Comment (xiongjiwei) [inline expression/column.go:43]: this change may cause the data race in `ParallelNestedLoopApplyExec`, which should fix later

**Review by xiongjiwei — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**